core/shared/src/main/scala/fs2/timeseries/TimeSeries.scala

```diff
@@ -0,0 +1,142 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+// Adapted from scodec-protocols, licensed under 3-clause BSD
+
+package fs2
+package timeseries
+
+import scala.concurrent.duration._
+
+import cats.effect.Temporal
+
+/** Companion for [[TimeSeries]]. */
+object TimeSeries {
+
+  /** Stream of either time ticks (spaced by `tickPeriod`) or values from the source stream. */
+  def apply[F[_]: Temporal, A](
+      source: Stream[F, TimeStamped[A]],
+      tickPeriod: FiniteDuration,
+      reorderOver: FiniteDuration
+  ): TimeSeries[F, A] = {
+    val src: TimeSeries[F, A] = source.map(tsa => tsa.map(Some(_): Option[A]))
+    val ticks: TimeSeries[F, Nothing] = timeTicks(tickPeriod).map(tsu => tsu.map(_ => None))
+    src.merge(ticks).through(TimeStamped.reorderLocally(reorderOver))
+  }
+
+  /** Stream of either time ticks (spaced by `tickPeriod`) or values from the source stream. */
+  def lift[F[_]: Temporal, A](
+      source: Stream[F, A],
+      tickPeriod: FiniteDuration,
+      reorderOver: FiniteDuration
+  ): TimeSeries[F, A] =
+    apply(source.map(TimeStamped.unsafeNow), tickPeriod, reorderOver)
+
+  /** Stream of time ticks spaced by `tickPeriod`. */
+  private def timeTicks[F[_]: Temporal](tickPeriod: FiniteDuration): Stream[F, TimeStamped[Unit]] =
+    Stream.awakeEvery[F](tickPeriod).map(_ => TimeStamped.unsafeNow(()))
+
+  /** Stream transducer that converts a stream of timestamped values with monotonically increasing timestamps in
+    * to a stream of timestamped ticks or values, where a tick is emitted every `tickPeriod`.
+    * Ticks are emitted between values from the source stream.
+    */
+  def interpolateTicks[A](
+      tickPeriod: FiniteDuration
+  ): Pipe[Pure, TimeStamped[A], TimeSeriesValue[A]] = {
+    def go(
+        nextTick: FiniteDuration,
+        s: Stream[Pure, TimeStamped[A]]
+    ): Pull[Pure, TimeSeriesValue[A], Unit] = {
+      def tickTime(x: Int) = nextTick + (x * tickPeriod)
+      s.pull.uncons.flatMap {
+        case Some((hd, tl)) =>
+          hd.indexWhere(_.time >= nextTick) match {
+            case None =>
+              if (hd.isEmpty) Pull.pure(())
+              else Pull.output(hd.map(_.toTimeSeriesValue)) >> go(nextTick, tl)
+            case Some(idx) =>
+              val (prefix, suffix) = hd.splitAt(idx)
+              val out =
+                if (prefix.isEmpty) Pull.pure(()) else Pull.output(prefix.map(_.toTimeSeriesValue))
```

Comment:
The `Pull.output` constructor should default to `Pull.done` if the chunk is empty. 
```suggestion
              val out = Pull.output(prefix.map(_.toTimeSeriesValue))
```